File: Downloads/project-bolt-github-7k4bwssw/project/server/routes/results.js
import express from 'express';
import multer from 'multer';
import ExcelJS from 'exceljs';
import PDFDocument from 'pdfkit';
import Response from '../models/Response.js';
import Exam from '../models/Exam.js';
import Student from '../models/Student.js';
import Solution from '../models/Solution.js';
import Result from '../models/Result.js';

const router = express.Router();
const upload = multer({ storage: multer.memoryStorage() });

// Process answer sheets
router.post('/:examId/process', upload.single('file'), async (req, res) => {
  try {
    const { examId } = req.params;
    
    if (!req.file) {
      return res.status(400).json({ error: 'No file uploaded' });
    }

    const exam = await Exam.findOne({ examId });
    if (!exam) {
      return res.status(404).json({ error: 'Exam not found' });
    }

    const solution = await Solution.findOne({ examId });
    if (!solution) {
      return res.status(400).json({ error: 'Solution not uploaded yet' });
    }

    const students = await Student.find({ examId });
    if (students.length === 0) {
      return res.status(400).json({ error: 'No students found' });
    }

    // For demo purposes, we'll simulate OCR processing
    // In a real implementation, you would use tesseract.js or similar
    const results = await simulateAnswerProcessing(exam, students, solution);
    
    // Save results to database
    await Result.deleteMany({ examId }); // Clear existing results
    await Result.insertMany(results);

    res.json({
      message: 'Answer sheets processed successfully',
      processedCount: results.length,
      results: results.map(r => ({
        studentName: r.studentName,
        score: r.score,
        totalMarks: r.totalMarks,
        percentage: r.percentage,
        result: r.result,
      })),
    });

  } catch (error) {
    console.error('Answer processing error:', error);
    res.status(500).json({ error: 'Failed to process answer sheets' });
  }
});

// Get results for an exam
router.get('/:examId/report', async (req, res) => {
  try {
    const { examId } = req.params;
    
    const exam = await Exam.findOne({ examId });
    if (!exam) {
      return res.status(404).json({ error: 'Exam not found' });
    }

    const results = await Result.find({ examId })
      .populate('studentId')
      .sort({ percentage: -1 });

    const reportData = results.map(result => ({
      studentName: result.studentId.name,
      lockerNumber: result.studentId.lockerNumber,
      rank: result.studentId.rank,
      score: result.score,
      totalMarks: result.totalMarks,
      percentage: result.percentage,
      result: result.result,
    }));

    const summary = {
      totalStudents: results.length,
      passCount: results.filter(r => r.result === 'Pass').length,
      failCount: results.filter(r => r.result === 'Fail').length,
      averagePercentage: results.reduce((sum, r) => sum + r.percentage, 0) / results.length,
      highestScore: Math.max(...results.map(r => r.score)),
      lowestScore: Math.min(...results.map(r => r.score)),
    };

    res.json({
      exam: {
        name: exam.name,
        dateTime: exam.dateTime,
        numQuestions: exam.numQuestions,
        marksPerMcq: exam.marksPerMcq,
        passingPercentage: exam.passingPercentage,
      },
      summary,
      results: reportData,
    });

  } catch (error) {
    console.error('Get results error:', error);
    res.status(500).json({ error: 'Failed to fetch results' });
  }
});

// Download results as Excel
router.get('/:examId/download/excel', async (req, res) => {
  try {
    const { examId } = req.params;
    
    const exam = await Exam.findOne({ examId });
    if (!exam) {
      return res.status(404).json({ error: 'Exam not found' });
    }

    const results = await Result.find({ examId })
      .populate('studentId')
      .sort({ percentage: -1 });

    const workbook = new ExcelJS.Workbook();
    const worksheet = workbook.addWorksheet('Results');

    // Headers
    worksheet.columns = [
      { header: 'Student Name', key: 'name', width: 25 },
      { header: 'Locker Number', key: 'lockerNumber', width: 15 },
      { header: 'Rank', key: 'rank', width: 15 },
      { header: 'Score', key: 'score', width: 10 },
      { header: 'Total Marks', key: 'totalMarks', width: 12 },
      { header: 'Percentage', key: 'percentage', width: 12 },
      { header: 'Result', key: 'result', width: 10 },
    ];

    // Add data
    results.forEach(result => {
      worksheet.addRow({
        name: result.studentId.name,
        lockerNumber: result.studentId.lockerNumber,
        rank: result.studentId.rank,
        score: result.score,
        totalMarks: result.totalMarks,
        percentage: `${result.percentage.toFixed(2)}%`,
        result: result.result,
      });
    });

    // Style headers
    worksheet.getRow(1).eachCell(cell => {
      cell.font = { bold: true };
      cell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFE6F3FF' } };
    });

    const buffer = await workbook.xlsx.writeBuffer();
    
    res.setHeader('Content-Type', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet');
    res.setHeader('Content-Disposition', `attachment; filename="${exam.name}_Results.xlsx"`);
    res.send(buffer);

  } catch (error) {
    console.error('Excel download error:', error);
    res.status(500).json({ error: 'Failed to download Excel report' });
  }
});

// Download results as PDF
router.get('/:examId/download/pdf', async (req, res) => {
  try {
    const { examId } = req.params;
    
    const exam = await Exam.findOne({ examId });
    if (!exam) {
      return res.status(404).json({ error: 'Exam not found' });
    }

    const results = await Result.find({ examId })
      .populate('studentId')
      .sort({ percentage: -1 });

    const doc = new PDFDocument({ margin: 50 });
    
    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="${exam.name}_Results.pdf"`);
    
    doc.pipe(res);

    // Title
    doc.fontSize(20).text('Exam Results Report', { align: 'center' });
    doc.fontSize(14).text(`Exam: ${exam.name}`, { align: 'center' });
    doc.text(`Date: ${exam.dateTime.toLocaleDateString()}`, { align: 'center' });
    doc.moveDown(2);

    // Summary
    const passCount = results.filter(r => r.result === 'Pass').length;
    const failCount = results.filter(r => r.result === 'Fail').length;
    const avgPercentage = results.reduce((sum, r) => sum + r.percentage, 0) / results.length;

    doc.fontSize(16).text('Summary', { underline: true });
    doc.fontSize(12);
    doc.text(`Total Students: ${results.length}`);
    doc.text(`Pass: ${passCount} | Fail: ${failCount}`);
    doc.text(`Average Percentage: ${avgPercentage.toFixed(2)}%`);
    doc.moveDown();

    // Results table
    doc.fontSize(14).text('Individual Results', { underline: true });
    doc.fontSize(10);
    
    const tableTop = doc.y + 20;
    doc.text('Name', 50, tableTop);
    doc.text('Locker #', 200, tableTop);
    doc.text('Rank', 270, tableTop);
    doc.text('Score', 320, tableTop);
    doc.text('Percentage', 370, tableTop);
    doc.text('Result', 450, tableTop);

    let currentY = tableTop + 20;
    results.forEach(result => {
      if (currentY > 700) {
        doc.addPage();
        currentY = 50;
      }
      
      doc.text(result.studentId.name, 50, currentY);
      doc.text(result.studentId.lockerNumber, 200, currentY);
      doc.text(result.studentId.rank, 270, currentY);
      doc.text(`${result.score}/${result.totalMarks}`, 320, currentY);
      doc.text(`${result.percentage.toFixed(2)}%`, 370, currentY);
      doc.text(result.result, 450, currentY);
      
      currentY += 15;
    });

    doc.end();

  } catch (error) {
    console.error('PDF download error:', error);
    res.status(500).json({ error: 'Failed to download PDF report' });
  }
});

// Helper function to simulate answer processing
async function simulateAnswerProcessing(exam, students, solution) {
  const results = [];
  
  for (const student of students) {
    // Simulate random answers for demo
    const answers = [];
    let correctCount = 0;
    
    for (let i = 1; i <= exam.numQuestions; i++) {
      const correctAnswer = solution.solutions.find(s => s.question === i)?.answer;
      const studentAnswer = Math.random() > 0.3 ? // 70% chance of answering
        ['A', 'B', 'C', 'D', 'E'][Math.floor(Math.random() * 5)] : 'BLANK';
      
      answers.push({ question: i, answer: studentAnswer });
      
      if (studentAnswer === correctAnswer) {
        correctCount++;
      }
    }
    
    const score = correctCount * exam.marksPerMcq;
    const totalMarks = exam.numQuestions * exam.marksPerMcq;
    const percentage = (score / totalMarks) * 100;
    const result = percentage >= exam.passingPercentage ? 'Pass' : 'Fail';
    
    results.push({
      examId: exam.examId,
      studentId: student._id,
      studentName: student.name,
      answers,
      score,
      totalMarks,
      percentage,
      result,
    });
  }
  
  return results;
}
// Get results for specific exam
router.get('/exam/:examId', async (req, res) => {
  try {
    const { examId } = req.params;
    const { page = 1, limit = 20, sortBy = 'processedAt', order = 'desc' } = req.query;

    const exam = await Exam.findOne({ examId });
    if (!exam) {
      return res.status(404).json({ error: 'Exam not found' });
    }

    const sortOrder = order === 'desc' ? -1 : 1;
    const skip = (page - 1) * limit;

    const responses = await Response.find({ examId })
      .sort({ [sortBy]: sortOrder })
      .skip(skip)
      .limit(parseInt(limit));

    const total = await Response.countDocuments({ examId });

    // Calculate aggregate statistics
    const stats = await calculateExamStats(examId);

    res.json({
      responses,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total,
        pages: Math.ceil(total / limit),
      },
      statistics: stats,
    });

  } catch (error) {
    console.error('Get results error:', error);
    res.status(500).json({ error: 'Failed to fetch results' });
  }
});

// Get individual student result
router.get('/student/:examId/:studentId', async (req, res) => {
  try {
    const { examId, studentId } = req.params;

    const response = await Response.findOne({ examId, studentId });
    if (!response) {
      return res.status(404).json({ error: 'Student response not found' });
    }

    const exam = await Exam.findOne({ examId });
    if (!exam) {
      return res.status(404).json({ error: 'Exam not found' });
    }

    // Provide detailed question-by-question analysis
    const questionAnalysis = response.responses.map((studentAnswer, index) => ({
      questionNumber: index + 1,
      studentAnswer,
      correctAnswer: exam.answerKey[index],
      isCorrect: studentAnswer === exam.answerKey[index],
      status: getAnswerStatus(studentAnswer),
    }));

    res.json({
      response,
      exam: {
        name: exam.name,
        numQuestions: exam.numQuestions,
      },
      questionAnalysis,
    });

  } catch (error) {
    console.error('Get student result error:', error);
    res.status(500).json({ error: 'Failed to fetch student result' });
  }
});

// Get aggregate statistics
router.get('/stats/:examId', async (req, res) => {
  try {
    const { examId } = req.params;
    const stats = await calculateExamStats(examId);
    res.json(stats);
  } catch (error) {
    console.error('Get stats error:', error);
    res.status(500).json({ error: 'Failed to fetch statistics' });
  }
});

// Helper function to calculate exam statistics
async function calculateExamStats(examId) {
  const responses = await Response.find({ examId });
  const exam = await Exam.findOne({ examId });

  if (responses.length === 0) {
    return {
      totalStudents: 0,
      averageScore: 0,
      highestScore: 0,
      lowestScore: 0,
      passingRate: 0,
      scoreDistribution: [],
      questionAnalysis: [],
    };
  }

  const scores = responses.map(r => r.score);
  const totalStudents = responses.length;
  const averageScore = scores.reduce((a, b) => a + b, 0) / totalStudents;
  const highestScore = Math.max(...scores);
  const lowestScore = Math.min(...scores);
  const passingScore = exam.settings?.passingScore || 60;
  const passingCount = scores.filter(score => (score / exam.numQuestions) * 100 >= passingScore).length;
  const passingRate = (passingCount / totalStudents) * 100;

  // Score distribution
  const ranges = [
    { min: 0, max: 20, label: '0-20%' },
    { min: 21, max: 40, label: '21-40%' },
    { min: 41, max: 60, label: '41-60%' },
    { min: 61, max: 80, label: '61-80%' },
    { min: 81, max: 100, label: '81-100%' },
  ];

  const scoreDistribution = ranges.map(range => {
    const count = scores.filter(score => {
      const percentage = (score / exam.numQuestions) * 100;
      return percentage >= range.min && percentage <= range.max;
    }).length;

    return {
      range: range.label,
      count,
      percentage: (count / totalStudents) * 100,
    };
  });

  // Question analysis
  const questionAnalysis = [];
  for (let i = 0; i < exam.numQuestions; i++) {
    const correctCount = responses.filter(r => r.responses[i] === exam.answerKey[i]).length;
    const correctPercentage = (correctCount / totalStudents) * 100;
    
    let difficulty = 'Easy';
    if (correctPercentage < 30) difficulty = 'Very Hard';
    else if (correctPercentage < 50) difficulty = 'Hard';
    else if (correctPercentage < 70) difficulty = 'Medium';

    // Find most common wrong answer
    const wrongAnswers = responses
      .map(r => r.responses[i])
      .filter(answer => answer !== exam.answerKey[i] && answer !== 'BLANK')
      .reduce((acc, answer) => {
        acc[answer] = (acc[answer] || 0) + 1;
        return acc;
      }, {});

    const commonWrongAnswer = Object.keys(wrongAnswers).length > 0
      ? Object.keys(wrongAnswers).reduce((a, b) => wrongAnswers[a] > wrongAnswers[b] ? a : b)
      : 'None';

    questionAnalysis.push({
      questionNumber: i + 1,
      correctPercentage: Math.round(correctPercentage),
      difficulty,
      commonWrongAnswer,
    });
  }

  return {
    totalStudents,
    averageScore: Math.round(averageScore * 100) / 100,
    highestScore,
    lowestScore,
    passingRate: Math.round(passingRate * 100) / 100,
    scoreDistribution,
    questionAnalysis,
  };
}

// Helper function to get answer status
function getAnswerStatus(answer) {
  switch (answer) {
    case 'BLANK': return 'No Answer';
    case 'MULTIPLE': return 'Multiple Marks';
    default: return 'Answered';
  }
}

export default router;

File: Downloads/FinalProductOmr/project/pServer/routers/results.py
from fastapi import APIRouter, HTTPException, Depends, UploadFile, File
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import List, Optional
import io
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter
from reportlab.lib.utils import ImageReader
from PIL import Image
from datetime import datetime
from .result import ResultCreate, ResultResponse
import logging
import json

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

router = APIRouter()

def get_database():
    from main import app
    return app.state.database

class BatchPDFRequest(BaseModel):
    examId: str
    examName: str
    results: List[dict]

@router.post("/save")
async def save_result(result: ResultCreate, db=Depends(get_database)):
    try:
        logger.info(f"Received result for saving: {result.dict()}")
        result_data = result.dict()
        result_data["processedAt"] = datetime.utcnow()

        existing_result = await db.results.find_one({
            "examId": result.examId,
            "studentId": result.studentId
        })

        if existing_result:
            await db.results.update_one(
                {"examId": result.examId, "studentId": result.studentId},
                {"$set": result_data}
            )
            logger.info(f"Updated existing result for student {result.studentId}")
        else:
            await db.results.insert_one(result_data)
            logger.info(f"Inserted new result for student {result.studentId}")

        return {"message": "Result saved successfully"}
    except ValueError as ve:
        logger.error(f"Validation error saving result: {str(ve)}")
        raise HTTPException(status_code=422, detail=f"Invalid input data: {str(ve)}")
    except Exception as e:
        logger.error(f"Failed to save result: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to save result: {str(e)}")

@router.post("/publish")
async def publish_results(publish_data: dict, db=Depends(get_database)):
    try:
        logger.info(f"Received publish request: {json.dumps(publish_data, default=str)}")
        exam_id = publish_data.get("examId")
        exam_name = publish_data.get("examName")
        results = publish_data.get("results", [])
        
        logger.info(f"Publishing {len(results)} results for exam {exam_id}")

        if not exam_id or not results:
            logger.error("Missing examId or results in publish request")
            raise HTTPException(status_code=400, detail="examId and results are required")

        exam = await db.exams.find_one({"examId": exam_id})
        if not exam:
            logger.error(f"Exam {exam_id} not found")
            raise HTTPException(status_code=404, detail="Exam not found")

        for result in results:
            result_data = {
                "examId": exam_id,
                "examName": exam_name,
                "studentId": result.get("studentId"),
                "studentName": result.get("studentName"),
                "rank": result.get("rank"),
                "lockerNumber": result.get("lockerNumber"),
                "score": result.get("score"),
                "totalMarks": result.get("totalMarks"),
                "percentage": result.get("percentage"),
                "passFailStatus": result.get("passFailStatus"),
                "correctAnswers": result.get("correctAnswers"),
                "incorrectAnswers": result.get("incorrectAnswers"),
                "blankAnswers": result.get("blankAnswers"),
                "multipleMarks": result.get("multipleMarks"),
                "responses": result.get("responses"),
                "sponsorDS": result.get("sponsorDS"),
                "course": result.get("course"),
                "wing": result.get("wing"),
                "module": result.get("module"),
                "publishedAt": datetime.utcnow()
            }

            existing_result = await db.results.find_one({
                "examId": exam_id,
                "studentId": result.get("studentId")
            })

            if existing_result:
                await db.results.update_one(
                    {"examId": exam_id, "studentId": result.get("studentId")},
                    {"$set": result_data}
                )
                logger.info(f"Updated published result for student {result.get('studentId')}")
            else:
                await db.results.insert_one(result_data)
                logger.info(f"Inserted published result for student {result.get('studentId')}")

        return {"message": f"Successfully published {len(results)} results for exam {exam_id}"}
    except Exception as e:
        logger.error(f"Failed to publish results: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to publish results: {str(e)}")

@router.get("/all", response_model=List[dict])
async def get_all_results(db=Depends(get_database)):
    try:
        logger.info("Fetching all results")
        cursor = db.results.find().sort("processedAt", -1)
        results = await cursor.to_list(length=None)
        
        for result in results:
            result['_id'] = str(result['_id'])
        
        logger.info(f"Retrieved {len(results)} results")
        return results
    except Exception as e:
        logger.error(f"Failed to fetch results: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to fetch results")

@router.post("/download-batch-pdf")
async def download_batch_pdf(request_data: BatchPDFRequest, images: List[UploadFile] = File(...)):
    try:
        logger.info(f"Received batch PDF request: {request_data.dict()}")
        logger.info(f"Received {len(images)} images")
        exam_id = request_data.examId
        exam_name = request_data.examName
        results = request_data.results
        
        logger.info(f"Generating batch PDF for exam {exam_id} with {len(results)} results")

        if not exam_id or not exam_name or not results:
            logger.error("Missing examId, examName, or results in request")
            raise HTTPException(status_code=400, detail="examId, examName, and results are required")

        if len(images) != len(results):
            logger.error(f"Number of images ({len(images)}) does not match number of results ({len(results)})")
            raise HTTPException(status_code=400, detail=f"Number of images: {len(images)} must match number of results ({len(results)})")
        
        required_fields = ["studentId", "studentName", "score", "totalMarks", "passFailStatus"]
        for result in results:
            missing_fields = [field for field in required_fields if field not in result or result[field] is None]
            if missing_fields:
                logger.error(f"Result missing required fields: {missing_fields}")
                raise HTTPException(status_code=400, detail=f"Result missing required fields: {missing_fields}")
        
        buffer = io.BytesIO()
        p = canvas.Canvas(buffer, pagesize=letter)
        width, height = letter
        margin = 50
        line_height = 20
        
        # Title page
        p.setFont("Helvetica-Bold", 18)
        y_pos = height - margin
        p.drawCentredString(width/2, y_pos, "OMR Checked Answer Sheets")
        p.setFont("Helvetica", 12)
        y_pos -= line_height
        p.drawCentredString(width/2, y_pos, f"Exam: {exam_name}")
        y_pos -= line_height
        p.drawCentredString(width/2, y_pos, f"Generated: {datetime.now().strftime('%Y-%m-%d')}")
        p.showPage()
        
        # Process each image and result
        for i, (image, result) in enumerate(zip(images, results)):
            if i > 0:
                p.showPage()
            
            # Reset position for new page
            y_pos = height - margin
            
            # Header
            p.setFont("Helvetica-Bold", 16)
            p.drawString(margin, y_pos, f"Answer Sheet - {result['studentName']}")
            y_pos -= line_height
            
            p.setFont("Helvetica", 12)
            p.drawString(margin, y_pos, f"Student ID: {result['studentId']}")
            y_pos -= line_height
            p.drawString(margin, y_pos, f"Rank: {result.get('rank', 'N/A')}")
            y_pos -= line_height
            p.drawString(margin, y_pos, f"Locker: {result.get('lockerNumber', 'N/A')}")
            y_pos -= line_height * 2
            
            # Read image data
            image_data = await image.read()
            img = Image.open(io.BytesIO(image_data))
            
            # Resize image to fit page while maintaining aspect ratio
            img_width, img_height = img.size
            aspect = img_height / float(img_width)
            target_width = width - (2 * margin)
            target_height = target_width * aspect
            
            if target_height > height - (y_pos + margin + 100):
                target_height = height - (y_pos + margin + 100)
                target_width = target_height / aspect
            
            # Draw image
            p.drawImage(ImageReader(io.BytesIO(image_data)), margin, margin + 50, target_width, target_height)
            y_pos = margin + 50 - line_height
            
            # Footer with result
            p.setFont("Helvetica-Bold", 12)
            p.setFillColorRGB(1, 0, 0)  # Red color
            result_text = f"Result: {result['score']}/{result['totalMarks']} - {result['passFailStatus']}"
            p.drawCentredString(width/2, margin, result_text)
            p.setFillColorRGB(0, 0, 0)  # Reset to black
        
        p.save()
        buffer.seek(0)
        
        logger.info(f"Generated batch PDF for exam {exam_id}")
        return StreamingResponse(
            io.BytesIO(buffer.read()),
            media_type="application/pdf",
            headers={"Content-Disposition": f"attachment; filename={exam_name}_Checked_Sheets.pdf"}
        )
    
    except Exception as e:
        logger.error(f"Failed to generate PDF: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to generate PDF: {str(e)}")

@router.post("/download-all-pdf")
async def download_all_pdf(request_data: dict):
    try:
        logger.info(f"Received summary PDF request: {json.dumps(request_data, default=str)}")
        results = request_data.get("results", [])
        filters = request_data.get("filters", {})
        
        logger.info(f"Generating summary PDF with {len(results)} results")

        buffer = io.BytesIO()
        p = canvas.Canvas(buffer, pagesize=letter)
        width, height = letter
        margin = 50
        line_height = 18
        y_pos = height - margin

        # Title
        p.setFont("Helvetica-Bold", 18)
        p.drawCentredString(width/2, y_pos, "OMR Results Report")
        y_pos -= line_height * 1.5
        p.setFont("Helvetica", 12)
        p.drawCentredString(width/2, y_pos, f"Generated: {datetime.now().strftime('%Y-%m-%d')}")
        y_pos -= line_height * 2

        # Summary Statistics
        total_students = len(results)
        passed_students = len([r for r in results if r.get('passFailStatus') == 'Pass'])
        failed_students = total_students - passed_students
        avg_percentage = sum(r.get('percentage', 0) for r in results) / total_students if total_students > 0 else 0

        p.setFont("Helvetica-Bold", 14)
        p.drawString(margin, y_pos, "Summary Statistics")
        y_pos -= line_height * 1.5
        p.setFont("Helvetica", 11)
        
        # Draw each statistic on a new line with increased spacing
        p.drawString(margin, y_pos, f"Total Students: {total_students}")
        y_pos -= line_height
        p.drawString(margin, y_pos, f"Passed: {passed_students} ({(passed_students/total_students*100):.1f}%)")
        y_pos -= line_height
        p.drawString(margin, y_pos, f"Failed: {failed_students} ({(failed_students/total_students*100):.1f}%)")
        y_pos -= line_height
        p.drawString(margin, y_pos, f"Average Score: {avg_percentage:.1f}%")
        y_pos -= line_height * 2

        # Results Table Header
        p.setFont("Helvetica-Bold", 9)
        headers = ["Student Name", "ID", "Exam", "Score", "%", "Result"]
        # Adjusted x-positions to prevent overlap
        x_positions = [margin, margin + 120, margin + 200, margin + 280, margin + 340, margin + 380]
        for header, x_pos in zip(headers, x_positions):
            p.drawString(x_pos, y_pos, header)
        y_pos -= line_height * 0.8
        p.line(margin, y_pos, width - margin, y_pos)  # Header underline
        y_pos -= line_height * 0.8

        # Individual Results
        p.setFont("Helvetica", 8)
        for result in results:
            if y_pos < margin + 50:  # Check for page break
                p.showPage()
                y_pos = height - margin
                # Redraw table header on new page
                p.setFont("Helvetica-Bold", 9)
                for header, x_pos in zip(headers, x_positions):
                    p.drawString(x_pos, y_pos, header)
                y_pos -= line_height * 0.8
                p.line(margin, y_pos, width - margin, y_pos)
                y_pos -= line_height * 0.8
                p.setFont("Helvetica", 8)

            # Draw row data with truncation to prevent overflow
            student_name = str(result.get('studentName', ''))[:25]  # Truncate long names
            exam_name = str(result.get('examName', ''))[:15]  # Truncate long exam names
            p.drawString(x_positions[0], y_pos, student_name)
            p.drawString(x_positions[1], y_pos, str(result.get('studentId', '')))
            p.drawString(x_positions[2], y_pos, exam_name)
            p.drawString(x_positions[3], y_pos, f"{result.get('score', 0)}/{result.get('totalMarks', 0)}")
            p.drawString(x_positions[4], y_pos, f"{result.get('percentage', 0):.1f}%")
            
            if result.get('passFailStatus') == 'Pass':
                p.setFillColorRGB(0, 0.5, 0)
            else:
                p.setFillColorRGB(1, 0, 0)
            p.drawString(x_positions[5], y_pos, str(result.get('passFailStatus', '')))
            p.setFillColorRGB(0, 0, 0)
            
            y_pos -= line_height * 0.9

        p.save()
        buffer.seek(0)

        logger.info("Generated summary PDF")
        return StreamingResponse(
            io.BytesIO(buffer.read()),
            media_type="application/pdf",
            headers={"Content-Disposition": "attachment; filename=OMR_Results_Report.pdf"}
        )

    except Exception as e:
        logger.error(f"Failed to generate summary PDF: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to generate PDF")

@router.get("/exam/{exam_id}")
async def get_exam_results(
    exam_id: str,
    page: int = 1,
    limit: int = 20,
    sort_by: str = "processedAt",
    order: str = "desc",
    db=Depends(get_database)
):
    try:
        logger.info(f"Fetching results for exam {exam_id}")
        
        query = {"examId": exam_id} if exam_id else {}
        
        if exam_id:
            exam = await db.exams.find_one({"examId": exam_id})
            if not exam:
                logger.error(f"Exam {exam_id} not found")
                raise HTTPException(status_code=404, detail="Exam not found")

        sort_order = -1 if order == "desc" else 1
        skip = (page - 1) * limit

        cursor = db.results.find(query).sort(sort_by, sort_order).skip(skip).limit(limit)
        responses = await cursor.to_list(length=None)

        total = await db.results.count_documents(query)

        for response in responses:
            response['_id'] = str(response['_id'])

        stats = await calculate_exam_stats(exam_id, db) if exam_id else {}

        logger.info(f"Retrieved {len(responses)} results" + (f" for exam {exam_id}" if exam_id else ""))
        return {
            "responses": responses,
            "pagination": {
                "page": page,
                "limit": limit,
                "total": total,
                "pages": (total + limit - 1) // limit
            },
            "statistics": stats
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to fetch results: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to fetch results")

async def calculate_exam_stats(exam_id: str, db):
    try:
        logger.info(f"Calculating stats for exam {exam_id}")
        cursor = db.results.find({"examId": exam_id})
        responses = await cursor.to_list(length=None)
        
        exam = await db.exams.find_one({"examId": exam_id})

        if not responses:
            logger.info(f"No results found for exam {exam_id}")
            return {
                "totalStudents": 0,
                "averageScore": 0,
                "highestScore": 0,
                "lowestScore": 0,
                "passingRate": 0,
                "scoreDistribution": [],
                "questionAnalysis": []
            }

        scores = [r["score"] for r in responses]
        total_students = len(responses)
        average_score = sum(scores) / total_students if total_students > 0 else 0
        highest_score = max(scores) if scores else 0
        lowest_score = min(scores) if scores else 0
        
        passing_score = exam.get("settings", {}).get("passingScore", 60)
        passing_count = len([s for s in scores if (s / exam["numQuestions"]) * 100 >= passing_score])
        passing_rate = (passing_count / total_students) * 100 if total_students > 0 else 0

        ranges = [
            {"min": 0, "max": 20, "label": "0-20%"},
            {"min": 21, "max": 40, "label": "21-40%"},
            {"min": 41, "max": 60, "label": "41-60%"},
            {"min": 61, "max": 80, "label": "61-80%"},
            {"min": 81, "max": 100, "label": "81-100%"}
        ]

        score_distribution = []
        for range_item in ranges:
            count = len([s for s in scores if range_item["min"] <= (s / exam["numQuestions"]) * 100 <= range_item["max"]])
            score_distribution.append({
                "range": range_item["label"],
                "count": count,
                "percentage": (count / total_students) * 100 if total_students > 0 else 0
            })

        logger.info(f"Calculated stats for exam {exam_id}: {total_students} students, {passing_rate:.1f}% passing")
        return {
            "totalStudents": total_students,
            "averageScore": round(average_score, 2),
            "highestScore": highest_score,
            "lowestScore": lowest_score,
            "passingRate": round(passing_rate, 2),
            "scoreDistribution": score_distribution,
            "questionAnalysis": []
        }
    except Exception as e:
        logger.error(f"Failed to calculate stats for exam {exam_id}: {str(e)}")
        raise